File: app/src/main/java/com/zack/enderplan/view/adapter/SingleTypePlanListAdapter.java
package com.zack.enderplan.view.adapter;

import android.content.res.ColorStateList;
import android.os.Handler;
import android.support.annotation.IntDef;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.zack.enderplan.R;
import com.zack.enderplan.util.ResourceUtil;
import com.zack.enderplan.util.StringUtil;
import com.zack.enderplan.util.TimeUtil;
import com.zack.enderplan.model.bean.Plan;
import com.zack.enderplan.view.widget.ImageTextView;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SingleTypePlanListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_ITEM = 1;
    private static final int TYPE_FOOTER = 2;

    public static final int PAYLOAD_CONTENT = 0;
    public static final int PAYLOAD_DEADLINE = 1;
    public static final int PAYLOAD_REMINDER = 2;
    public static final int PAYLOAD_STAR = 3;

    @IntDef({SCROLL_EDGE_TOP, SCROLL_EDGE_MIDDLE, SCROLL_EDGE_BOTTOM})
    @Retention(RetentionPolicy.SOURCE)
    private @interface ScrollEdge {}

    public static final int SCROLL_EDGE_TOP = -1;
    public static final int SCROLL_EDGE_MIDDLE = 0;
    public static final int SCROLL_EDGE_BOTTOM = 1;

    private List<Plan> mSingleTypePlanList;
    private OnStarStatusChangedListener mOnStarStatusChangedListener;
    private OnPlanItemClickListener mOnPlanItemClickListener;

    private int mAccentColor, mGrey600Color;

    private int mScrollEdge;

    public SingleTypePlanListAdapter(List<Plan> singleTypePlanList) {
        mSingleTypePlanList = singleTypePlanList;

        mAccentColor = ResourceUtil.getColor(R.color.colorAccent);
        mGrey600Color = ResourceUtil.getColor(R.color.grey_600);

        mScrollEdge = SCROLL_EDGE_TOP;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case TYPE_ITEM:
                return new ItemViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_list_single_type_plan, parent, false));
            case TYPE_FOOTER:
                return new FooterViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.footer_list_single_type_plan, parent, false));
            default:
                throw new IllegalArgumentException("The argument viewType cannot be " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (getItemViewType(position)) {
            case TYPE_ITEM:
                ItemViewHolder itemViewHolder = (ItemViewHolder) holder;

                Plan plan = mSingleTypePlanList.get(position);

                setContentText(itemViewHolder.mContentText, plan.getContent(), plan.isCompleted());
                setSpaceView(itemViewHolder.mSpaceView, plan.isCompleted(), plan.hasDeadline(), plan.hasReminder());
                setTimeLayout(itemViewHolder.mDeadlineLayout, plan.isCompleted(), plan.hasDeadline(), plan.getDeadline());
                setTimeLayout(itemViewHolder.mReminderLayout, plan.isCompleted(), plan.hasReminder(), plan.getReminderTime());
                setStarButton(itemViewHolder.mStarButton, plan.isStarred(), plan.isCompleted(), itemViewHolder);
                setItemView(itemViewHolder);
                break;
            case TYPE_FOOTER:
                FooterViewHolder footerViewHolder = (FooterViewHolder) holder;
                setSingleTypePlanCountText(footerViewHolder.mSingleTypePlanCountText);
                break;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position, List<Object> payloads) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position);
        } else {
            ItemViewHolder itemViewHolder = (ItemViewHolder) holder;
            Plan plan = mSingleTypePlanList.get(position);
            for (Object payload : payloads) {
                switch ((int) payload) {
                    case PAYLOAD_CONTENT:
                        setContentText(itemViewHolder.mContentText, plan.getContent(), plan.isCompleted());
                        break;
                    case PAYLOAD_DEADLINE:
                        setSpaceView(itemViewHolder.mSpaceView, plan.isCompleted(), plan.hasDeadline(), plan.hasReminder());
                        setTimeLayout(itemViewHolder.mDeadlineLayout, plan.isCompleted(), plan.hasDeadline(), plan.getDeadline());
                        break;
                    case PAYLOAD_REMINDER:
                        setSpaceView(itemViewHolder.mSpaceView, plan.isCompleted(), plan.hasDeadline(), plan.hasReminder());
                        setTimeLayout(itemViewHolder.mReminderLayout, plan.isCompleted(), plan.hasReminder(), plan.getReminderTime());
                        break;
                    case PAYLOAD_STAR:
                        setStarButton(itemViewHolder.mStarButton, plan.isStarred(), plan.isCompleted(), itemViewHolder);
                        break;
                }
            }
        }
    }

    @Override
    public int getItemCount() {
        return mSingleTypePlanList.size() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        return position == mSingleTypePlanList.size() ? TYPE_FOOTER : TYPE_ITEM;
    }

    public void notifyItemInsertedAndChangingFooter(int position) {
        notifyItemInserted(position);
        notifyFooterChanged();
    }

    public void notifyItemRemovedAndChangingFooter(int position) {
        notifyItemRemoved(position);
        notifyFooterChanged();
    }

    public void notifyListScrolled(@ScrollEdge int scrollEdge) {
        if (mScrollEdge == scrollEdge) return;
        int lastScrollEdge = mScrollEdge;
        mScrollEdge = scrollEdge;
        if (lastScrollEdge == SCROLL_EDGE_BOTTOM || mScrollEdge == SCROLL_EDGE_BOTTOM) {
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    notifyFooterChanged();
                }
            }, 10);
        }
    }

    private void setContentText(TextView contentText, String content, boolean isCompleted) {
        contentText.setText(isCompleted ? StringUtil.addSpan(content, StringUtil.SPAN_STRIKETHROUGH) : content);
    }

    private void setSpaceView(View spaceView, boolean isCompleted, boolean hasDeadline, boolean hasReminder) {
        spaceView.setVisibility(isCompleted || (!hasDeadline && !hasReminder) ? View.GONE : View.VISIBLE);
    }

    private void setTimeLayout(ImageTextView timeLayout, boolean isCompleted, boolean hasTime, long time) {
        timeLayout.setVisibility(!isCompleted && hasTime ? View.VISIBLE : View.GONE);
        timeLayout.setText(hasTime ? TimeUtil.formatTime(time) : null);
    }

    private void setStarButton(final ImageView starButton, boolean isStarred, boolean isCompleted, final ItemViewHolder itemViewHolder) {
        setStarButtonImage(starButton, isStarred, isCompleted);
        starButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int layoutPosition = itemViewHolder.getLayoutPosition();
                if (mOnStarStatusChangedListener != null) {
                    mOnStarStatusChangedListener.onStarStatusChanged(layoutPosition);
                }
                Plan plan = mSingleTypePlanList.get(layoutPosition);
                setStarButtonImage(starButton, plan.isStarred(), plan.isCompleted());
            }
        });
    }

    private void setStarButtonImage(ImageView starButton, boolean isStarred, boolean isCompleted) {
        starButton.setImageResource(isStarred ? R.drawable.ic_star_black_24dp : R.drawable.ic_star_border_black_24dp);
        starButton.setImageTintList(ColorStateList.valueOf(!isStarred || isCompleted ? mGrey600Color : mAccentColor));
    }

    private void setItemView(final ItemViewHolder itemViewHolder) {
        if (mOnPlanItemClickListener != null) {
            itemViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mOnPlanItemClickListener.onPlanItemClick(itemViewHolder.getLayoutPosition());
                }
            });
        }
    }

    private void setSingleTypePlanCountText(TextView singleTypePlanCountText) {
        singleTypePlanCountText.setVisibility(mScrollEdge == SCROLL_EDGE_BOTTOM ? View.VISIBLE : View.INVISIBLE);
        singleTypePlanCountText.setText(ResourceUtil.getQuantityString(R.plurals.text_plan_count, mSingleTypePlanList.size()));
    }

    private void notifyFooterChanged() {
        notifyItemChanged(mSingleTypePlanList.size());
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.text_content)
        TextView mContentText;
        @BindView(R.id.view_space)
        View mSpaceView;
        @BindView(R.id.layout_deadline)
        ImageTextView mDeadlineLayout;
        @BindView(R.id.layout_reminder)
        ImageTextView mReminderLayout;
        @BindView(R.id.btn_star)
        ImageView mStarButton;

        public ItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    class FooterViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.text_single_type_plan_count)
        TextView mSingleTypePlanCountText;

        public FooterViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public interface OnPlanItemClickListener {
        void onPlanItemClick(int position);
    }

    public void setOnPlanItemClickListener(OnPlanItemClickListener listener) {
        mOnPlanItemClickListener = listener;
    }

    public interface OnStarStatusChangedListener {
        void onStarStatusChanged(int position);
    }

    public void setOnStarStatusChangedListener(OnStarStatusChangedListener listener) {
        mOnStarStatusChangedListener = listener;
    }
}

File: app/src/main/kotlin/me/imzack/app/end/view/adapter/SingleTypePlanListAdapter.kt
package me.imzack.app.end.view.adapter

import android.content.res.ColorStateList
import android.os.Handler
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import me.imzack.app.end.R
import me.imzack.app.end.common.Constant
import me.imzack.app.end.model.bean.Plan
import me.imzack.app.end.util.ResourceUtil
import me.imzack.app.end.util.StringUtil
import me.imzack.app.end.util.TimeUtil
import me.imzack.app.end.view.widget.ImageTextView

class SingleTypePlanListAdapter(private val mSingleTypePlanList: List<Plan>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    //TODO 可以把这两个属性放到构造函数中
    var mOnPlanItemClickListener: ((position: Int) -> Unit)? = null
    var mOnStarStatusChangedListener: ((position: Int) -> Unit)? = null

    private val mAccentColor = ResourceUtil.getColor(R.color.colorAccent)
    private val mGrey600Color = ResourceUtil.getColor(R.color.grey_600)

    private var mScrollEdge = Constant.SCROLL_EDGE_TOP

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            when (viewType) {
                Constant.VIEW_TYPE_ITEM -> ItemViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_list_single_type_plan, parent, false))
                Constant.VIEW_TYPE_FOOTER -> FooterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.footer_list_single_type_plan, parent, false))
                else -> throw IllegalArgumentException("The argument viewType cannot be " + viewType)
            }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            Constant.VIEW_TYPE_ITEM -> {
                val itemViewHolder = holder as ItemViewHolder

                val plan = mSingleTypePlanList[position]

                setContentText(itemViewHolder.mContentText, plan.content, plan.isCompleted)
                setSpaceView(itemViewHolder.mSpaceView, plan.isCompleted, plan.hasDeadline, plan.hasReminder)
                setTimeLayout(itemViewHolder.mDeadlineLayout, plan.isCompleted, plan.hasDeadline, plan.deadline)
                setTimeLayout(itemViewHolder.mReminderLayout, plan.isCompleted, plan.hasReminder, plan.reminderTime)
                setStarButton(itemViewHolder.mStarButton, plan.isStarred, plan.isCompleted, itemViewHolder)
                setItemView(itemViewHolder)
            }
            Constant.VIEW_TYPE_FOOTER -> {
                val footerViewHolder = holder as FooterViewHolder
                setSingleTypePlanCountText(footerViewHolder.mSingleTypePlanCountText)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int, payloads: List<Any>) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position)
        } else {
            val itemViewHolder = holder as ItemViewHolder
            val plan = mSingleTypePlanList[position]
            for (payload in payloads) {
                when (payload as Int) {
                    Constant.PLAN_PAYLOAD_CONTENT -> setContentText(itemViewHolder.mContentText, plan.content, plan.isCompleted)
                    Constant.PLAN_PAYLOAD_DEADLINE -> {
                        setSpaceView(itemViewHolder.mSpaceView, plan.isCompleted, plan.hasDeadline, plan.hasReminder)
                        setTimeLayout(itemViewHolder.mDeadlineLayout, plan.isCompleted, plan.hasDeadline, plan.deadline)
                    }
                    Constant.PLAN_PAYLOAD_REMINDER_TIME -> {
                        setSpaceView(itemViewHolder.mSpaceView, plan.isCompleted, plan.hasDeadline, plan.hasReminder)
                        setTimeLayout(itemViewHolder.mReminderLayout, plan.isCompleted, plan.hasReminder, plan.reminderTime)
                    }
                    Constant.PLAN_PAYLOAD_STAR_STATUS -> setStarButton(itemViewHolder.mStarButton, plan.isStarred, plan.isCompleted, itemViewHolder)
                }
            }
        }
    }

    override fun getItemCount() = mSingleTypePlanList.size + 1

    override fun getItemViewType(position: Int) = if (position == mSingleTypePlanList.size) Constant.VIEW_TYPE_FOOTER else Constant.VIEW_TYPE_ITEM

    fun notifyItemInsertedAndChangingFooter(position: Int) {
        notifyItemInserted(position)
        notifyFooterChanged()
    }

    fun notifyItemRemovedAndChangingFooter(position: Int) {
        notifyItemRemoved(position)
        notifyFooterChanged()
    }

    fun notifyListScrolled(scrollEdge: Int) {
        if (mScrollEdge == scrollEdge) return
        val lastScrollEdge = mScrollEdge
        mScrollEdge = scrollEdge
        if (lastScrollEdge == Constant.SCROLL_EDGE_BOTTOM || mScrollEdge == Constant.SCROLL_EDGE_BOTTOM) {
            Handler().postDelayed({ notifyFooterChanged() }, 10)
        }
    }

    private fun setContentText(contentText: TextView, content: String, isCompleted: Boolean) {
        contentText.text = if (isCompleted) StringUtil.addSpan(content, StringUtil.SPAN_STRIKETHROUGH) else content
    }

    private fun setSpaceView(spaceView: View, isCompleted: Boolean, hasDeadline: Boolean, hasReminder: Boolean) {
        spaceView.visibility = if (isCompleted || !hasDeadline && !hasReminder) View.GONE else View.VISIBLE
    }

    private fun setTimeLayout(timeLayout: ImageTextView, isCompleted: Boolean, hasTime: Boolean, time: Long) {
        timeLayout.visibility = if (!isCompleted && hasTime) View.VISIBLE else View.GONE
        timeLayout.mText = if (hasTime) TimeUtil.formatDateTime(time) else null
        timeLayout.updateText()
    }

    private fun setStarButton(starButton: ImageView, isStarred: Boolean, isCompleted: Boolean, itemViewHolder: ItemViewHolder) {
        setStarButtonImage(starButton, isStarred, isCompleted)
        starButton.setOnClickListener {
            val layoutPosition = itemViewHolder.layoutPosition
            mOnStarStatusChangedListener?.invoke(layoutPosition)
            val plan = mSingleTypePlanList[layoutPosition]
            setStarButtonImage(starButton, plan.isStarred, plan.isCompleted)
        }
    }

    private fun setStarButtonImage(starButton: ImageView, isStarred: Boolean, isCompleted: Boolean) {
        starButton.setImageResource(if (isStarred) R.drawable.ic_star_black_24dp else R.drawable.ic_star_border_black_24dp)
        starButton.imageTintList = ColorStateList.valueOf(if (!isStarred || isCompleted) mGrey600Color else mAccentColor)
    }

    private fun setItemView(itemViewHolder: ItemViewHolder) {
        itemViewHolder.itemView.setOnClickListener { mOnPlanItemClickListener?.invoke(itemViewHolder.layoutPosition) }
    }

    private fun setSingleTypePlanCountText(singleTypePlanCountText: TextView) {
        singleTypePlanCountText.visibility = if (mScrollEdge == Constant.SCROLL_EDGE_BOTTOM) View.VISIBLE else View.INVISIBLE
        singleTypePlanCountText.text = ResourceUtil.getQuantityString(R.plurals.text_plan_count, mSingleTypePlanList.size)
    }

    private fun notifyFooterChanged() {
        notifyItemChanged(mSingleTypePlanList.size)
    }

    class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.text_content)
        lateinit var mContentText: TextView
        @BindView(R.id.view_space)
        lateinit var mSpaceView: View
        @BindView(R.id.layout_deadline)
        lateinit var mDeadlineLayout: ImageTextView
        @BindView(R.id.layout_reminder)
        lateinit var mReminderLayout: ImageTextView
        @BindView(R.id.btn_star)
        lateinit var mStarButton: ImageView

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    class FooterViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.text_single_type_plan_count)
        lateinit var mSingleTypePlanCountText: TextView

        init {
            ButterKnife.bind(this, itemView)
        }
    }
}
